File: ProblemSet/464 · 整数排序 II/python/464 · 整数排序 II 堆排序.py
class Solution:
    """
    @param A: an integer array
    @return: nothing
    """
    def sortIntegers2(self, A):
        self.n = len(A)
        self.heapSort(A)

    def heapSort(self, nums):
        for i in range((self.n >> 1) - 1, -1, -1):
            self.maxHeapify(nums, i, self.n)
        for i in range(self.n - 1, 0, -1):
            nums[0], nums[i] = nums[i], nums[0]
            self.maxHeapify(nums, 0, i)

    def maxHeapify(self, nums, root, len):
        tmp = nums[root]
        while (root << 1) + 1 < len:
            child = (root << 1) + 1
            if child + 1 != len and nums[child] < nums[child + 1]:
                child += 1
            if tmp >= nums[child]:
                break
            else:
                nums[root] = nums[child]
            root = child
        nums[root] = tmp

File: ProblemSet/464 · 整数排序 II/java/464 · 整数排序 II 堆排序.java
public class Solution {
    /**
     * @param A: an integer array
     * @return: nothing
     */
    public void sortIntegers2(int[] A) {
        heapSort(A);
    }
<<<<<<< HEAD
    
=======

>>>>>>> 1e7c6ca88d99bb5c35ad853537c8c7954ef14ee3
    private void heapSort(int[] nums) {
        for (int i = (nums.length >> 1) - 1; i >= 0; --i)
            maxHeapify(nums, i, nums.length);
        for (int i = nums.length - 1; i > 0; --i) {
            swap(nums, 0, i);
            maxHeapify(nums, 0, i);
        }
    }

    private void maxHeapify(int[] nums, int root, int len) {
        int tmp = nums[root];
        while ((root << 1) + 1 < len) {
            int child = (root << 1) + 1;
            if (child + 1 != len && nums[child] < nums[child + 1]) ++child;
            if (tmp >= nums[child]) break;
            else nums[root] = nums[child];
            root = child;
        }
        nums[root] = tmp;
    }

    private void swap(int[] nums, int i, int j) {
        int tmp = nums[i];
        nums[i] = nums[j];
        nums[j] = tmp;
    }
}